File: library/src/main/java/com/senierr/permission/PermissionManager.java
package com.senierr.permission;

import android.app.Activity;
import android.app.FragmentManager;

/**
 * 权限管理
 *
 * @author zhouchunjie
 * @date 2017/3/6
 */
public class PermissionManager {

    private static final String TAG = PermissionManager.class.getSimpleName();

    private ProxyFragment checkFragment;
    private String[] permissions;

    private PermissionManager(Activity activity) {
        if (checkFragment == null) {
            checkFragment = (ProxyFragment) activity.getFragmentManager().findFragmentByTag(TAG);
            if (checkFragment == null) {
                checkFragment = new ProxyFragment();
                FragmentManager fragmentManager = activity.getFragmentManager();
                fragmentManager.beginTransaction()
                        .add(checkFragment, TAG)
                        .commitAllowingStateLoss();
                fragmentManager.executePendingTransactions();
            }
        }
    }

    /**
     * 创建
     *
     * @param activity
     * @return
     */
    public static PermissionManager with(Activity activity) {
        return new PermissionManager(activity);
    }

    /**
     * 添加权限
     *
     * @param permissions
     * @return
     */
    public PermissionManager permissions(final String... permissions) {
        if (permissions == null || permissions.length == 0) {
            throw new IllegalArgumentException("PermissionManager requires at least one input permission!");
        }
        this.permissions = permissions;
        return this;
    }

    /**
     * 检查权限
     *
     * @return
     */
    public boolean check() {
        if (checkFragment != null) {
            return checkFragment.check(permissions);
        } else {
            return false;
        }
    }

    /**
     * 请求权限
     *
     * @param checkCallback
     */
    public void request(RequestCallback checkCallback) {
        if (checkFragment != null) {
            checkFragment.setRequestCallback(checkCallback);
            checkFragment.request(permissions);
        }
    }
}

File: library/src/main/java/com/senierr/permission/PermissionManager.kt
package com.senierr.permission

import android.support.v4.app.Fragment
import android.support.v4.app.FragmentActivity

/**
 * 权限管理
 *
 * @author zhouchunjie
 * @date 2017/3/6
 */
class PermissionManager private constructor(activity: FragmentActivity) {

    companion object {

        private val TAG = PermissionManager::class.java.simpleName

        @JvmStatic
        fun with(activity: FragmentActivity): PermissionManager {
            return PermissionManager(activity)
        }

        @JvmStatic
        fun with(fragment: Fragment): PermissionManager? {
            val activity = fragment.activity
            return if (activity != null) PermissionManager(activity) else null
        }
    }

    private var proxyFragment: ProxyFragment? = null
    private var permissions: Array<out String>? = null

    init {
        proxyFragment = activity.supportFragmentManager.findFragmentByTag(TAG) as ProxyFragment?
        if (proxyFragment == null) {
            proxyFragment = ProxyFragment()
            proxyFragment?.let {
                val fragmentManager = activity.supportFragmentManager
                fragmentManager.beginTransaction()
                        .add(it, TAG)
                        .commitAllowingStateLoss()
                fragmentManager.executePendingTransactions()
            }
        }
    }

    /**
     * 添加权限
     */
    fun permissions(vararg permissions: String): PermissionManager {
        if (permissions.isEmpty()) throw IllegalArgumentException("PermissionManager requires at least one input permission!")
        this.permissions = permissions
        return this
    }

    /**
     * 检查权限
     */
    fun check() = proxyFragment?.check(permissions) ?: false

    /**
     * 请求权限
     */
    fun request(requestCallback: RequestCallback) {
        proxyFragment?.request(permissions, requestCallback)
    }
}